### PR TITLE
Secure Environment

### PR DESCRIPTION
**Pull request by mattmanning**
### Description
This PR is carrying forward https://github.com/convox/rack/pull/2120 opened by @ravenac95.

This change allows users to specify that a certain service should not have its environment variables injected into the ECS TaskDefinition, where they are visible in plaintext to users of the AWS account.

Instead, environment variables are injected into the TaskDefinition that identify the URL of the encrypted environment file and the ID of the KMS key that can decrypt them. It is up to the code in the application process itself to download the file and use the KMS key to decrypt the secrets.

### Summary for release notes

A secure environment can be specified for a container in the `docker-compose.yml` file by adding `convox.environment.secure=true` to the labels, like so:

```
version: "2"

services:
  web:
    build: .
    ports:
      - 4444:4444
    labels:
      - convox.environment.secure=true
```

Once the secure environment is enabled, 3 environment variables will be passed into the container at runtime. These variables are: 

* `SECURE_ENVIRONMENT_URL` - The url to the s3 file that contains the environment (encrypted with KMS)
* `SECURE_ENVIRONMENT_TYPE` - The type of the file that is encrypted in s3. This is reserved for the future. For now it only returns `envfile`
* `SECURE_ENVIRONMENT_KEY` - The arn to the KMS key. Used for decryption of the environment file referenced by  `SECURE_ENVIRONMENT_URL`

### Guidance for reviewers

This change can be tested by updating your Rack to a branch release built from this branch. The most recent is `20170414183923-ravenac95-add-secure-env`.

Add the `convox.environment.secure=true` label to a service and deploy the app. Once the app is deployed, verify that its environment variables no longer exist in the new TaskDefinition.

To verify that the "secure" variables are available, you can use the tool forked from @ravenac95's project here:  [convox/secure-environment](https://github.com/convox/secure-environment).

There is a simple example of using this tool with a Docker ENTRYPOINT here: [convox/secure-env-example](https://github.com/convox/secure-env-example)

### Before Release
- [ ] Rebase against master
- [ ] Get review approval
- [ ] Confirm test coverage
- [ ] Submit documentation PR for convox/site

### CODE REVIEW

**Comment by mattmanning** [comment]
[Use KMS to decrypt secrets within the app. Improve access controls to secrets stored in S3.](https://trello.com/c/kovHzAx5/2345-use-kms-to-decrypt-secrets-within-the-app-improve-access-controls-to-secrets-stored-in-s3)

File: manifest/service.go [hunk 1]
```diff
@@ -105,6 +105,14 @@ func (s *Service) Process(app string, m Manifest) Process {
 	return NewProcess(app, *s, m)
 }
 
+// UseSecureEnvironment - Determines if the service intends to use a secure environment
+func (s Service) UseSecureEnvironment() bool {
+	if useSecureEnvironment, ok := s.Labels["convox.environment.secure"]; useSecureEnvironment == "true" && ok {
```
Comment (MiguelMoll) [inline manifest/service.go]: nit: could this function be simplified to `return s.Labels["convox.environment.secure"] == "true"`?

File: manifest/service_test.go [hunk 2]
```diff
@@ -92,3 +92,15 @@ func TestSyncPaths(t *testing.T) {
 		}
 	}
 }
+
+func TestUseSecureEnvironment(t *testing.T) {
+	m, err := manifestFixture("secure-env")
+	if err != nil {
+		assert.FailNow(t, fmt.Sprintf("failed to read fixture: %s", err.Error()))
```
Comment (MiguelMoll) [inline manifest/service_test.go]: There's a `require.NoError()` function that wraps this error check logic with nice output.

File: provider/aws/lambda/formation/handler/ecs.go [hunk 3]
```diff
@@ -206,13 +211,41 @@ func ECSTaskDefinitionCreate(req Request) (string, map[string]string, error) {
...
+		secureEnv := false
+
+		if rawSecureEnv, ok := task["SecureEnvironment"].(string); ok {
+			secureEnv, err = strconv.ParseBool(rawSecureEnv)
+			if err != nil {
+				return "invalid", nil, err
+			}
+		}
+
+		if secureEnv && envURL != "" {
```
Comment (MiguelMoll) [inline provider/aws/lambda/formation/handler/ecs.go]: Second check for empty `envURL`?

**Comment by mwarkentin** [comment]
@mattmanning @MiguelMoll @ddollar @nzoschke One question - how does this fit in with the discussions about using the EC2 Parameter Store at some point for managing configs? Does this handle what we need, or would there still be a potential of moving to that at some point?

**Comment by mattmanning** [comment]
@mwarkentin We haven't ruled out EC2 parameter store afaik, but this was a quick way to piggyback on the existing env system to get some extra security.

**Pull request closed without merging**